go/libraries/doltcore/env/remotes.go

```diff
@@ -205,6 +207,9 @@ func NewPushOpts(ctx context.Context, apr *argparser.ArgParseResults, rsr RepoSt
 	switch src.GetType() {
 	case ref.BranchRefType:
 		remoteRef, err = GetTrackingRef(dest, remote)
+		if !setUpstream {
+			setUpstream = gotRemoteAndBranch
```

Comment:
I added a top-level comment for this, too, but wanted to mention it with the code, too. I don't think this is a behavior that Git performs (automatically setting upstream on push when remote and branch are specified), but let me know if you think I'm just missing something.